### PR TITLE
The sprints and multiple assignee

### PR DESCRIPTION
**Pull request by crz11111**

### CODE REVIEW

File: src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskModification.java [hunk 1]
```diff
@@ -24,12 +30,25 @@ public TaskModification(int task_id, Consumer<ITask> callback) {
...
+        final var assignees_sel = new MultiselectComboBox<User>();
+        assignees_sel.setItems(NitriteDBProvider.getInstance().getUsers().collect(Collectors.toList()));
+        assignees_sel.setPlaceholder("Assigning to..");
+        Set<User> currentAssignees = new HashSet<>();
+        task.getAssignees().forEach(assigneesId -> currentAssignees.add(NitriteDBProvider.getInstance().getUser(assigneesId)));
+        assignees_sel.setValue(currentAssignees);
+
+        final var sprints = new String[(int) NitriteDBProvider.getInstance().getSprints4Project(NitriteDBProvider.getInstance().getCurrentProject(CurrentUser.get())).count()];
+        for(int i=1; i<= sprints.length; i++){
+            sprints[i-1]= "Sprint " + i;
```
Comment (RuthgerD) [inline src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskModification.java]: "Maybe its better to list the sprint start date.."

File: src/main/java/org/coffeemine/app/spring/data/User.java [hunk 2]
```diff
@@ -113,4 +113,8 @@ public User fromNO2Doc(Document doc) {
         account_passhash = doc.get("account_passhash", String.class);
         return this;
     }
+
+    public String toString(){
+        return name;
```
Comment (RuthgerD) [inline src/main/java/org/coffeemine/app/spring/data/User.java]: this is just .getName();

File: src/main/java/org/coffeemine/app/spring/data/ITask.java [hunk 3]
```diff
@@ -13,6 +13,10 @@
 
     void setName(@NotNull String name);
 
+    @NotNull String getAssignSprint();
```
Comment (RuthgerD) [inline src/main/java/org/coffeemine/app/spring/data/ITask.java]: this reduced OOP design and its better to make a helper function in the database to search for you

File: src/main/java/org/coffeemine/app/spring/data/Task.java [hunk 4]
```diff
@@ -12,6 +12,8 @@
     @NotNull
     private String name = "";
     @NotNull
+    private String assignSprint = "";
```
Comment (RuthgerD) [inline src/main/java/org/coffeemine/app/spring/data/Task.java]: even if we want to continue using this you should at least store the sprint id integer and not string

File: src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskDetail.java [hunk 5]
```diff
@@ -18,7 +18,10 @@ public TaskDetail(int id, Consumer<ITask> on_edit) {
         final var name = new Text(task.getName());
         final var id_txt = new Text(Integer.toString(task.getId()));
         final var description = new Text(task.getDescription());
-        final var bound_sprint = new Text("Sprint X");
+        final var currentAssignees = new StringBuffer();
+        task.getAssignees().forEach(assigneesId -> currentAssignees.append(NitriteDBProvider.getInstance().getUser(assigneesId).toString()));
```
Comment (RuthgerD) [inline src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskDetail.java]: String.join(", ", task.getAssignees().stream().map(t -> db.getUser(t).getName()).toArray(size -> new String[size]));

File: src/main/java/org/coffeemine/app/spring/TestingData.java [hunk 6]
```diff
@@ -30,10 +30,10 @@ static void load() {
...
 
         var tasks = new ArrayList<Integer>();
-        tasks.add(db.addTask(new Task(-1, "Hello0", "Task1", 10, true, users, new ArrayList<>(), new ArrayList<>())));
-        tasks.add(db.addTask(new Task(-1, "Hello1", "Task2", 16, true, users, new ArrayList<>(), new ArrayList<>())));
-        tasks.add(db.addTask(new Task(-1, "Hello2", "Task3", 1, true, users, new ArrayList<>(), new ArrayList<>())));
-        tasks.add(db.addTask(new Task(-1, "Hello3", "Task4", 3, false, users, new ArrayList<>(), new ArrayList<>())));
+        tasks.add(db.addTask(new Task(-1, "Hello0", "Sprint 1", "Task1", 10, true, users, new ArrayList<>(), new ArrayList<>())));
+        tasks.add(db.addTask(new Task(-1, "Hello1", "Sprint 1", "Task2", 16, true, users, new ArrayList<>(), new ArrayList<>())));
+        tasks.add(db.addTask(new Task(-1, "Hello2", "Sprint 1", "Task3", 1, true, users, new ArrayList<>(), new ArrayList<>())));
+        tasks.add(db.addTask(new Task(-1, "Hello3", "Sprint 1", "Task4", 3, false, users, new ArrayList<>(), new ArrayList<>())));
```
[the remark this answers is not in the document]
Comment (AeroStun) [reply src/main/java/org/coffeemine/app/spring/TestingData.java]: And even if it was to be changed, do not identify sprints by name

File: src/main/java/org/coffeemine/app/spring/components/EventsDialog/SprintCreation.java [hunk 7]
```diff
@@ -19,13 +20,16 @@
     public SprintCreation(Consumer<ISprint> callback) {
         super();
 
-        final var title = new Text("Sprint" + " x");
+        final var title = new Text(
+                "Sprint " + ((int) NitriteDBProvider.getInstance().getSprints4Project(
+                                NitriteDBProvider.getInstance().getCurrentProject(CurrentUser.get())).count()+1));
```
[the remark this answers is not in the document]
Comment (AeroStun) [reply src/main/java/org/coffeemine/app/spring/components/EventsDialog/SprintCreation.java]: Definitely; use the starting date

File: src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskCreation.java [hunk 8]
```diff
@@ -7,29 +7,46 @@
...
         sprint_sel.setPlaceholder("Assigning to sprint..");
+
         final var create_btn = new Button("Create", event -> {
-            task.setName(name.getValue());
-            task.setDescription(desc.getValue());
-            NitriteDBProvider.getInstance().addTask(task);
+            ArrayList<Integer> usersId = new ArrayList<>();
+            assignees_sel.getSelectedItems().forEach(user -> usersId.add(user.getId()));
+            task = NitriteDBProvider.getInstance().getTask(
+                    NitriteDBProvider.getInstance().addTask(
```
[the remark this answers is not in the document]
Comment (AeroStun) [reply src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskCreation.java]: More readable and more efficient

File: src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskModification.java [hunk 9]
```diff
@@ -7,10 +7,16 @@
 import com.vaadin.flow.component.select.Select;
 import com.vaadin.flow.component.textfield.TextArea;
 import com.vaadin.flow.component.textfield.TextField;
+import org.coffeemine.app.spring.auth.CurrentUser;
 import org.coffeemine.app.spring.data.ITask;
+import org.coffeemine.app.spring.data.User;
 import org.coffeemine.app.spring.db.NitriteDBProvider;
+import org.vaadin.gatanaso.MultiselectComboBox;
 
+import java.lang.reflect.Array;
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskModification.java]: Pretty sure this is an unused import

File: src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskModification.java [hunk 10]
```diff
@@ -24,12 +30,25 @@ public TaskModification(int task_id, Consumer<ITask> callback) {
...
+        assignees_sel.setItems(NitriteDBProvider.getInstance().getUsers().collect(Collectors.toList()));
+        assignees_sel.setPlaceholder("Assigning to..");
+        Set<User> currentAssignees = new HashSet<>();
+        task.getAssignees().forEach(assigneesId -> currentAssignees.add(NitriteDBProvider.getInstance().getUser(assigneesId)));
+        assignees_sel.setValue(currentAssignees);
+
+        final var sprints = new String[(int) NitriteDBProvider.getInstance().getSprints4Project(NitriteDBProvider.getInstance().getCurrentProject(CurrentUser.get())).count()];
+        for(int i=1; i<= sprints.length; i++){
+            sprints[i-1]= "Sprint " + i;
+        }
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskModification.java]: Naked for-loops are 99% of the time a red flag indicating that you're not using abstractions well enough

File: src/main/java/org/coffeemine/app/spring/data/ITask.java (hunk 3 above)
Comment (AeroStun) [reply src/main/java/org/coffeemine/app/spring/data/ITask.java]: +1

File: src/main/java/org/coffeemine/app/spring/data/Task.java (hunk 4 above)
Comment (AeroStun) [reply src/main/java/org/coffeemine/app/spring/data/Task.java]: We shouldn't backtrack sprints from tasks

File: src/main/java/org/coffeemine/app/spring/data/User.java (hunk 2 above)
Comment (AeroStun) [reply src/main/java/org/coffeemine/app/spring/data/User.java]: +1

Comment (crz11111) [reply src/main/java/org/coffeemine/app/spring/data/User.java]: No really. Cause the multi-select box needs toString() to show the object name.

Comment (AeroStun) [reply src/main/java/org/coffeemine/app/spring/data/User.java]: Then use an inline proxy class

File: src/main/java/org/coffeemine/app/spring/calendar/Calendar.java [hunk 11]
```diff
@@ -88,7 +88,7 @@ void updateEvents() {
                     sprint.getEnd().plusDays(1).atStartOfDay(),
                     true,
                     true,
-                    "red",
+                    "Green",
```
[the remark this answers is not in the document]
Comment (crz11111) [reply src/main/java/org/coffeemine/app/spring/calendar/Calendar.java]: There you go

File: src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskDetail.java [hunk 12]
```diff
@@ -18,9 +18,7 @@ public TaskDetail(int id, Consumer<ITask> on_edit) {
         final var name = new Text(task.getName());
         final var id_txt = new Text(Integer.toString(task.getId()));
         final var description = new Text(task.getDescription());
-        final var currentAssignees = new StringBuffer();
-        task.getAssignees().forEach(assigneesId -> currentAssignees.append(NitriteDBProvider.getInstance().getUser(assigneesId).toString()));
-        final var bound_assignees =  new Text(currentAssignees.toString());
+        final var bound_assignees =  new Text(String.join(", ", task.getAssignees().stream().map(t -> NitriteDBProvider.getInstance().getUser(t).getName()).toArray(String[]::new)));
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/components/EventsDialog/TaskDetail.java]: Could use a reduce

**Pull request merged**